### PR TITLE
fix: Avoid data races for cacheEntries and index files

### PR DESCRIPTION
**Pull request by cbuto**
Closes https://github.com/helm/chartmuseum/issues/546.

This should resolve the data races seen when running the test suite with the `-race` flag.

I took some of the work/feedback from https://github.com/helm/chartmuseum/pull/549 (thanks @nerdeveloper!) and added some additional changes to ensure we are locking when access the index file and cache entries.

I still would like to spend more time on the performance aspect to see if adjusting the locks could help. 

Overall this change as is has a slight negative effect on median response time and a extremely positive effect the p95 response time. I assume the p95 response without these changes is fairly high due to data races between requests. 

---

Here are the results from the **very** small load tests I ran locally.

1. Start chartmuseum locally
```
rm -rf /tmp/chartmuseum-loadtest-*  
make build-mac && ./bin/darwin/amd64/chartmuseum --storage local --storage-local-rootdir /tmp
```

2. Run locust with 10 users/10 spawn rate

Here are the test run branches:

Test 8 = `main`
Test 9 = `fix/cache-race`
Test 10 = `main`
Test 11 = `fix/cache-race`

Screen Shot 2022-03-04 at 6 12 57 PM

### CODE REVIEW

**Comment by scbizu** [comment]
wow , looks good . 

I will spend some more time to read over this PR 👀   

However , it will conflict with #549 ?

**Comment by cbuto** [comment]
> wow , looks good .
> 
> I will spend some more time to read over this PR 👀
> 
> However , it will conflict with #549 ?

Thanks! No rush at all. But yeah builds off of #549 and the initial work @nerdeveloper did.

**Comment by scbizu** [comment]
I read some of this code , and have a question: what we do is actually make the rough-grain lock (RegenerationLock) to be the fine-grain lock (RepoLock &  IndexLock) ? But do we need to keep these rough-grain lock after we did this ?  

And I see the response time should be reduced literally , and this  optimization also related to the #396 .

**Comment by cbuto** [comment]
> I read some of this code , and have a question: what we do is actually make the rough-grain lock (RegenerationLock) to be the fine-grain lock (RepoLock & IndexLock) ? But do we need to keep these rough-grain lock after we did this ?
> 
> And I see the response time should be reduced literally , and this optimization also related to the #396 .

good call! I think you are right, I removed the RegenerationLock in 48b75c77920e87d100eafa37d40f4bc049c1ee7d and it looks good. Some quick load test results showed no change but i don't think that lock was necessary anymore!

**Comment by scbizu** [comment]
Looks much more comfortable now , I will review these all changes and make some tests on my mac this weekend .

**Review by scbizu — CHANGES_REQUESTED** [review]
The code here should be really careful for both writing and reviewing XD

**Comment by scbizu** [comment]
@jdolitsky hihi , Could you also join the reviewing , I think it is an important fix or refactor PR .

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 1]
```diff
@@ -682,8 +685,11 @@ func (suite *MultiTenantServerTestSuite) TestOverwriteServer() {
 		// only for testing purpose
 		time.Sleep(time.Second)
 		// depth: 0
-		e := suite.extractRepoEntryFromInternalCache("")
+		e, err := suite.extractRepoEntryFromInternalCache("")
+		suite.Nil(err, "no error extracting repo entry from internal cache")
+		e.RepoLock.Lock()
```
[the remark this answers is not in the document]
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server_test.go]: yep! missed that, changed it in 3619fdf7cf34529fafc8c59e227195c9c571b80b

File: pkg/chartmuseum/server/multitenant/cache.go [hunk 2]
```diff
@@ -620,10 +625,26 @@ func (server *MultiTenantServer) refreshCacheEntry(log cm_logger.LoggingFn, repo
...
 	entry.RepoIndex = ir.index
-
 	if server.UseStatefiles {
 		// Dont wait, save index-cache.yaml to storage in the background.
 		// It is not crucial if this does not succeed, we will just log any errors
 		go server.saveStatefile(log, repo, ir.index.Raw)
 	}
 }
+
+func (m *memoryCacheStore) Load(key interface{}) (*cacheEntry, bool, error) {
```
[the remark this answers is not in the document]
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/cache.go]: yep definitely, its a bit weird returning an error. Fixed it in 3619fdf7cf34529fafc8c59e227195c9c571b80b

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 3]
```diff
@@ -657,12 +657,15 @@ func (suite *MultiTenantServerTestSuite) TestDisabledDeleteServer() {
...
+	local, ok, err := suite.OverwriteServer.InternalCacheStore.Load(repo)
+	if err != nil {
+		return nil, err
+	}
 	if ok {
-		return local
+		return local, nil
 	}
-	return nil
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server_test.go]: after fixing https://github.com/helm/chartmuseum/pull/556#discussion_r824878627, this return value will go away

**Review by scbizu — APPROVED** [review]
LGTM

**Review by nerdeveloper — APPROVED** [review]

**Comment by nerdeveloper** [comment]
Awesome work there @cbuto

**Comment by cbuto** [comment]
after rebasing with https://github.com/helm/chartmuseum/pull/551, some new data races showed up while running the tests (it seems to be flakey)

**Review by jdolitsky — APPROVED** [review]
fantastic. thanks for addressing this (apologies for delay in review!)

**Comment by cbuto** [comment]
latest race fixed in 39464355182051cc2b5ebedd19bd46d5899dbfed, I ran the tests quite a few times just to be sure and it looks good!

**Pull request merged**